core/src/main/scala/org/apache/spark/rdd/UnionRDD.scala

```diff
@@ -62,7 +64,23 @@ class UnionRDD[T: ClassTag](
     var rdds: Seq[RDD[T]])
   extends RDD[T](sc, Nil) {  // Nil since we implement getDependencies
 
+  // Evaluate partitions in parallel. Partitions of each rdd will be cached by the `partitions`
+  // val in `RDD`.
+  private[spark] lazy val parallelPartitionEval: Boolean = {
```

Comment:
Yeah I was referring to the side point about caching, but I likely misunderstood. Yes getSplits is the issue. HadoopRDD has an InputFormat and it is possible the driver calls getPartitions on it simultaneously because two actions could be invoked in parallel. This is the basic scenario I am thinking of though it has a bunch of variants.